### ISSUE TITLE
JIT: Add missing new-style ABI classifiers for LA64 and RISCV64

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/runtime#100744]
#100138 introduced a new representation for ABI information that I am planning on switching the JIT to use in all places that deal with ABI details. It explicitly represents how all parts of all parameters are passed. #100276 and #100526 introduced classifiers for all of our supported targets: win-x86, win-x64, SysV x64 and arm64.
#100572 is an example PR that is making use of the new ABI representation to rewrite parameter homing.

Before I can proceed with moving the rest of the JIT to the new representation we need implementations of new-style classifiers for LA64 (cc @shushanhf) and RISCV64 (cc @dotnet/samsung). I would greatly appreciate contributions of implementations of these. Otherwise I can try to base it on what happens in `lvaInitUserArgs` today, but I will not be able to test it.

Once implemented, the following code can be enabled to cross check the new style ABI information against the old one stored in `LclVarDsc`: https://github.com/dotnet/runtime/blob/9b57a265c7efd3732b035bade005561a04767128/src/coreclr/jit/lclvars.cpp#L1809-L1892
Eventually the old information will be removed from `LclVarDsc` and the new style information will be the only source-of-truth for the ABI information.

**Comment by shushanhf** [comment]
Thanks, I will study these new code.

**Comment by tomeksowi** [comment]
> @bartlomiejko I have to get off work early. So can someone in your team help RISCV64 implementation and tests?

I can do that once I'm done with #100080, it's a similar area.

**Comment by tomeksowi** [comment]
@jakobbotsch If a sub-word value is passed, shouldn't ABIPassingSegment also contain info whether the unused bits are (sign|zero)-extended (or NaN-boxed) vs. undefined?

**Comment by jakobbotsch** [comment]
> @jakobbotsch If a sub-word value is passed, shouldn't ABIPassingSegment also contain info whether the unused bits are (sign|zero)-extended (or NaN-boxed) vs. undefined?

Do you think it would be useful? ARM32 has similar requirements in its ABI, but I don't immediately see where we would end up actually querying this information on `ABIPassingSegment` today.
I would probably wait with adding it until we see some more evidence of it being useful, or do some work to make use of it. Maybe that will show up naturally once we move more things to the new representation.

Another missing piece right now is information about when structs are passed by reference, but I'm also delaying adding that information until we move the things that actually query this information to the new representation.

**Comment by tomeksowi** [comment]
> Do you think it would be useful? ARM32 has similar requirements in its ABI, but I don't immediately see where we would end up actually querying this information on `ABIPassingSegment` today.

There's a bit of a discrepancy where e.g. an integer primitive is (sign|zero)-extended but for struct { float; int; } the unused integer bits are unspecified. But it could also be inferred by looking at whether the type is a struct or not, so it's more of a convenience than must-have.

> I would probably wait with adding it until we see some more evidence of it being useful, or do some work to make use of it. Maybe that will show up naturally once we move more things to the new representation.

Makes sense, thanks.

**Comment by shushanhf** [comment]
@jakobbotsch 
The new-style ABI is used to replace the `Compiler::lvaInitUserArgs{}` where resolve the args info, is it?
If we add the classifier for LA64/RV64, we will delete the `Compiler::lvaInitUserArgs{}` ?

**Comment by jakobbotsch** [comment]
> @jakobbotsch The new-style ABI is used to replace the `Compiler::lvaInitUserArgs{}` where resolve the args info, is it? If we add the classifier for LA64/RV64, we will delete the `Compiler::lvaInitUserArgs{}` ? IMO, I think we can uniform the classifier and the `Compiler::lvaInitUserArgs{}`, that is, can we cache the classifier info within the `Compiler::lvaInitUserArgs{}` ？

Yes, we will eventually remove the ABI code from `lvaInitUserArgs` once everything has moved. The function will stay and be responsible for creating and initializing `LclVarDsc`, but without any ABI related code.

There is no need to do the ABI classification at the same time. I looked at throughput when introducing the separation and the impact is minimal. Separating the concerns simplifies the code because we initialize parameters (like this pointer and ret buffer) from multiple places, not just `lvaInitUserArgs`.

**Comment by shushanhf** [comment]
> > @jakobbotsch The new-style ABI is used to replace the `Compiler::lvaInitUserArgs{}` where resolve the args info, is it? If we add the classifier for LA64/RV64, we will delete the `Compiler::lvaInitUserArgs{}` ? IMO, I think we can uniform the classifier and the `Compiler::lvaInitUserArgs{}`, that is, can we cache the classifier info within the `Compiler::lvaInitUserArgs{}` ？
> 
> Yes, we will eventually remove the ABI code from `lvaInitUserArgs` once everything has moved. The function will stay and be responsible for creating and initializing `LclVarDsc`, but without any ABI related code.
> 
> There is no need to do the ABI classification at the same time. I looked at throughput when introducing the separation and the impact is minimal. Separating the concerns simplifies the code because we initialize parameters (like this pointer and ret buffer) from multiple places, not just `lvaInitUserArgs`.
> 
> Eventually my goal is to also use this for `GenTreeCall` ABI info.

OK, I see.
I will add the `LoongArch64Classifier::Classify` within the `targetloongarch64.cpp`.

**Comment by tomeksowi** [comment]
> but RISCV64 still has its own version. @tomeksowi do you plan to submit a PR to switch it?

Yes, I'm working on it
